### ISSUE TITLE
unhandled stderr.writeln() in default mode catch block

### ISSUE THREAD

**Issue by psahabu** [chapel-lang/chapel#7130]
### Summary of Problem
Calling `stderr.writeln()` from a catch block in default mode causes incomprehensible errors to be emitted, rather than a properly inserted `halt()`. Explicitly wrapping the call in a `try!` preempts the problem.

### Steps to Reproduce

**Source Code:**

```chapel
proc unrelated() {
  try {
    writeln("inside try");
  } catch {
    stderr.writeln("stderr");
  }
}

unrelated();
```

**Compile command:**
`chpl unrelated.chpl`

**Execution command:**
`./unrelated`

### Configuration Information

- Output of `chpl --version`: 1.16 prerelease
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
  - CHPL_TARGET_PLATFORM: darwin
  - CHPL_TARGET_COMPILER: clang
  - CHPL_TARGET_ARCH: native
  - CHPL_LOCALE_MODEL: flat
  - CHPL_COMM: none
  - CHPL_TASKS: qthreads
  - CHPL_LAUNCHER: none
  - CHPL_TIMERS: generic
  - CHPL_UNWIND: none
  - CHPL_MEM: jemalloc
  - CHPL_MAKE: make
  - CHPL_ATOMICS: intrinsics
  - CHPL_GMP: gmp
  - CHPL_HWLOC: hwloc
  - CHPL_REGEXP: re2
  - CHPL_WIDE_POINTERS: struct
  - CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
  - Apple LLVM version 8.0.0 (clang-800.0.42.1)
  - Target: x86_64-apple-darwin15.6.0
  - Thread model: posix
  - InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Comment by psahabu** [comment]
Ah, this is the band-aid fix in #6607 coming back to bite me, `errorHandling.cpp:337`:

``` cpp
Expr* insert = node->getStmtExpr();
if (insert == NULL)
  insert = node;
```

Here's the AST that is given to the `lowerErrorHandling` pass from `resolve`:

```
val _return_tmp_[1201912]:bool[7] "temp"
(1201914 'move' _return_tmp_[1201912](193214 call writeln[1074430] stderr[235915] "stderr"))
```

I'm actually not sure this is normalized correctly, but in any case `getStmtExpr()` returns NULL while inspecting the stderr CallExpr, causing the insert point to be the stderr CallExpr itself. Here's the AST that gets output:

```
val _return_tmp_[1201912]:bool[7] "temp"
(1201914 'move' _return_tmp_[1201912] val error[1231757]:Error[176388] "error variable" "temp"(1231759 'move' error[1231757] nil[34])(193214 call writeln[1074430] stderr[235915] "stderr" error[1231757])
```

Any ideas on how to move forward @mppf?

**Comment by mppf** [comment]
@psahabu - I don't see anything wrong with the input to lowerErrorHandling - it looks normalized OK to me.

Ah, I see that node->parentExpr == NULL when error handling enterCallExpr handles it.
I set a watchpoint to find when it was set to NULL (it wasn't this way after resolution, or --verify would fail). It is set to NULL on line 237 (`node->replace(tryBody)`) Then it is set to something else on line 323 (`info.tryBody->insertAtTail`).

Recall that we decided that the visiting order is:
  * Try
      * enter try statement
      * exit try statement
      * for each catch statement
          * enter catch statement
          * exit catch statement

This is the problematic pattern:
 * exit try statement, which runs `node->replace(tryBody);`, thereby removing the catch bodies from the AST tree (since they are inside the try statement).
 * when handling the inside of the catch statement, in enterCallExpr, the contents of the catch statement are not currently in the AST tree, and so the call in question isn't in the tree
 * in exitTryStmt, in lowerCatches info.tryBody->insertAtTail, the handler (including the user code) is once again added to the tree

I'll let you come up with a solution here. One "obvious" strategy would be to track the parent expression in our visitor (it's always the first CallExpr inside of a BlockStmt). I'm sure there is a more subtle but perhaps better strategy that would allow the catch body to remain in the tree.

**Comment by psahabu** [comment]
Fixed by #7269, a nice PR.
